modules/apps/29-fee/keeper/escrow.go

```diff
@@ -50,50 +50,76 @@ func (k Keeper) EscrowPacketFee(ctx sdk.Context, packetID channeltypes.PacketId,
 }
 
 // DistributePacketFees pays the acknowledgement fee & receive fee for a given packetID while refunding the timeout fee to the refund account associated with the Fee.
-func (k Keeper) DistributePacketFees(ctx sdk.Context, forwardRelayer string, reverseRelayer sdk.AccAddress, feesInEscrow []types.PacketFee) {
+func (k Keeper) DistributePacketFees(ctx sdk.Context, forwardRelayer string, reverseRelayer sdk.AccAddress, packetFees []types.PacketFee, isTimeout bool) {
```

Comment:
The refactor here is great, I think it's a massive improvement splitting up the timeout and ack logic as you have. I can't help but feel like I'd prefer to have two separate functions though rather than passing a boolean. I feel like the fact that we're adding a boolean here could be a sign this [fn has too much responsibility](https://en.wikipedia.org/wiki/Single-responsibility_principle). 

What about the idea of having `DistributeAcknowledgementPacketFees` & `DistributeTimeoutPacketFees`?

I'm open to either way, but just thought I'd mention. Wdyt? :)

Reply:
Yeah, I agree. I think it might be worth exploring the option of two separate functions in favour of the boolean arg, considering that the forward relayer is always an empty string for timeout scenarios also. 

I do appreciate that it would be a decent bit of duplicate code nonetheless, but I would lean towards having explicit APIs for each scenario (happy path/timeout)

Reply:
My main concern is the sensitivity of performing the cacheCtx. It can be hard to verify that we pass in `cacheCtx` and `ctx` in the correct functions. Doing this redundantly in two separate functions seems like a recipe for one of the functions to get out of sync with the other and potentially lead to a bug. 

The actual distribution of fees on acknowledgement and fees on timeout are still separated. The logic this function is performing is wrapping packet fee distribution with an escrow out of balance check. The previously functionality not only distributed the packet fees for either ack or timeout but also wrapped it with an escrow out of balance check. 

For reference you can see how this file [differed before splitting](https://github.com/cosmos/ibc-go/blob/4d5a7474ca971e6aeb6b1dab55727b66353ed361/modules/apps/29-fee/keeper/escrow.go#L53-#L141). I actually find it harder to verify the [`distributeAcknowledgementPacketFees`](https://github.com/cosmos/ibc-go/blob/4d5a7474ca971e6aeb6b1dab55727b66353ed361/modules/apps/29-fee/keeper/escrow.go#L73-#L91) and [`distirbuteTimeoutPacketFees`](https://github.com/cosmos/ibc-go/blob/4d5a7474ca971e6aeb6b1dab55727b66353ed361/modules/apps/29-fee/keeper/escrow.go#L121-#L134) functionality in comparison to the proposed way [`distributePacketFeeOnAcknowledgement`](https://github.com/cosmos/ibc-go/blob/317aaa191ee92f64549694fad8bb0660272ccf14/modules/apps/29-fee/keeper/escrow.go#L95-#L109) and [`distirbutePacketFeeOnTimeout`](https://github.com/cosmos/ibc-go/blob/317aaa191ee92f64549694fad8bb0660272ccf14/modules/apps/29-fee/keeper/escrow.go#L114-#L122)

Reply:
I agree with @seantking and @damiannolan. I think having two separate functions will make the code more explicit and easier to read.

If that's not possible because of the concerns that @colin-axner has regarding the context, then a couple of other alternatives could be:
1. To have an enum value instead of a boolean as parameter. Then at least reading the function at call site would look like `..., OnTimeout)` or `..., OnAcknowledgement)`.
2. To pass the function that actually does the distribution instead of a boolean as parameter. So at call site you would have something like `..., distributePacketFeeOnAcknowledgement)` or `..., distributePacketFeeOnTimeout)`. But then you need to have both functions with the same signature, which is possible to work around, but might not lead to the most beautiful result...